### ISSUE TITLE
Visual C++ "Smaller Type Check"

### ISSUE THREAD

**Issue by kenhagin** [fmtlib/fmt#3683]
I use Visual C++ 2022/C++20.  When enabling Smaller Type Check in Debug configurations, I have found that format.h triggers the error in 4 places (that I've hit).  This is not an error per-se, but I found it was easily eliminated by masking uint64 with UINT32_MAX before the cast, preventing the code from seeing high bits (intentionally) discarded by the cast.  Example:

line 3449: prod = static_cast<uint32_t>(prod & UINT32_MAX) * static_cast<uint64_t>(100); // I inserted the UINT32_MAX

The 4 locations I hit and modified are on or near the following lines:  3449, 3476, 3505, and 3516.

I believe modifying these should not break the code on any other platform (unless UINT32_MAX isn't available everywhere) and will support use of "Smaller Type Check" in VC.

**Comment by vitaut** [comment]
Thanks for the suggestion but I think it just obscures the logic without brining any benefits. In general fixing all possible warnings is a non-goal.

**Comment by kenhagin** [comment]
Thank you.  However, I would like to point out that “Smaller Type Check” does not produce a compile time warning of any kind.  It causes a runtime debug trap that breaks execution.  This means if I want to use {fmt} (it is great!) without editing after each update, I have to turn off “Smaller Type Check” which I would like to use to expose bugs in my main program.  That’s why I reported it.  I do agree that squelching odd warnings should not be a goal.

 

I have been using it as “header only”.  I suppose I could turn it into a library and disable “Smaller Type Check” for that library project.  It was just so easy to use “header only” as I’m only using it in one routine that is called millions of times when dumping output, and using FMT_COMPILE cuts 1/3 of my runtime.  I’ll look into that.

 

Thanks,

Ken Hagin

----------------------------------------

Home e-mail: ***@***.***

**Comment by mwinterb** [comment]
Doesn't the Microsoft STL not support "Smaller Type Check"?
https://developercommunity.visualstudio.com/t/meow/859581#T-N778441

That check still seems to be in current code:
https://github.com/microsoft/STL/blob/1c59a205787f4fdcd9c20decc8fa6dfa9096d394/stl/inc/yvals_core.h#L1933-L1938

**Comment by kenhagin** [comment]
My solution has precompiled headers “pch.h”, and in my lowest level library that is a dependency for everything else I have a “pchcommon.h” that all projects’ pch.h includes.  _ALLOW_RTCc_IN_STL is the first thing defined.  I have not trapped on it within any STL code yet.

 

Meanwhile, I assume that it is working because it trapped in format.h.  I did debug the code and figured out that in a 64-bit unsigned there is a high 32-bit value and a low 32-bit value, and the casting traps because the high 32-bit value is non-zero.  I could tell this was by design, not broken code.

 

To keep things simple, I want to continue to use {fmt} as header_only, so I can just alter the 4 places I found the issue after each minor update.  It’s a small price to pay.

 

Thanks,

Ken Hagin

----------------------------------------

Home e-mail: ***@***.***
